lib/rust/metamodel/src/meta/sexp.rs

```diff
@@ -0,0 +1,181 @@
+        }
+        for (i, field) in fields.iter().enumerate() {
+            if let Some(name) = field.name.to_camel_case() {
+                let car = Value::Symbol(name.into_boxed_str());
+                let cdr = self.value(field.type_, data);
+                out.push(Value::Cons(Cons::new(car, cdr)));
+            } else {
+                out.push(self.value(field.type_, data));
+            }
+            if self.graph[id].child_field == Some(i + 1) {
+                self.fields(hierarchy, data, out);
+            }
+        }
+    }
+
+    fn primitive(&self, primitive: Primitive, data: &mut &[u8]) -> Value {
+        match primitive {
+            Primitive::U32 => Value::Number(read_u32(data).into()),
+            Primitive::U64 => Value::Number(read_u64(data).into()),
+            Primitive::Bool => {
+                let value = read_u8(data);
+                let value = match value {
+                    0 => false,
+                    1 => true,
+                    _ => panic!(),
+                };
+                Value::Bool(value)
+            }
+            Primitive::String => Value::String(read_string(data).into()),
+            Primitive::Sequence(t0) => {
+                let len = read_u64(data);
+                Value::vector((0..len).map(|_| self.value(t0, data)))
+            }
+            Primitive::Option(t0) => match read_u8(data) {
+                0 => Value::Null,
+                1 => self.value(t0, data),
+                _ => panic!(),
+            },
+            Primitive::Result(t0, t1) => {
+                let mut values = vec![];
+                match read_u32(data) {
+                    0 => {
+                        values.push(Value::Symbol("Ok".to_owned().into_boxed_str()));
+                        values.push(self.value(t0, data));
+                    }
+                    1 => {
+                        values.push(Value::Symbol("Err".to_owned().into_boxed_str()));
+                        values.push(self.value(t1, data));
+                    }
+                    _ => panic!(),
+                }
+                Value::list(values)
+            }
+        }
+    }
+}
+
+
+// === Primitive Deserializers ===
+
+fn read_u8(buffer: &mut &[u8]) -> u8 {
+    let (bytes, rest) = buffer.split_at(1);
+    *buffer = rest;
+    bytes[0]
+}
+
+fn read_u32(buffer: &mut &[u8]) -> u32 {
+    let (bytes, rest) = buffer.split_at(4);
+    *buffer = rest;
+    let mut data = [0; 4];
+    data.copy_from_slice(bytes);
+    u32::from_le_bytes(data)
+}
+
+fn read_u64(buffer: &mut &[u8]) -> u64 {
+    let (bytes, rest) = buffer.split_at(8);
+    *buffer = rest;
+    let mut data = [0; 8];
+    data.copy_from_slice(bytes);
+    u64::from_le_bytes(data)
+}
+
+fn read_string(buffer: &mut &[u8]) -> String {
+    let len = read_u64(buffer);
+    let (bytes, rest) = buffer.split_at(len as usize);
+    *buffer = rest;
+    String::from_utf8(bytes.to_owned()).unwrap()
+}
```

Comment:
could you add some tests to the file to show the usage scenario here, please?

Reply:
Let's create a practical use case test and link to it from this file then! :)

Reply:
Upon reflection (ha), I've moved S-expression support to a separate crate. It is not a core part of metamodel.